### PR TITLE
Refactor transaction build command

### PR DESCRIPTION
**Pull request by Jimbo4350**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Convenience/Query.hs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                    case eEraHistory of
+                      Left e -> return $ Left $ AcqFailure e
+                      Right eraHistory ->
+                        case eSystemStart of
+                          Left e -> return $ Left $ AcqFailure e
+                          Right systemStart ->
+                            case eStakePools of
+                              Left e -> return $ Left e
+                              Right stakePools ->
+                               return $ Right (utxo, pparams, eraHistory, systemStart, stakePools)
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/Convenience/Query.hs:112]: Personally, I'd prefer do-notation here

```suggestion
            let mapLeft f = either (Left . f) Right
            return $ do
              utxo <- eUtxo
              pparam <- ePparams
              eraHistory <- mapLeft AcqFailure eEraHistory
              systemStart <- mapLeft AcqFailure eSystemStart
              stakePools <- eStakePools
              Right (utxo, pparams, eraHistory, systemStart, stakePools)
```

File: cardano-api/src/Cardano/Api/Convenience/Construction.hs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+module Cardano.Api.Convenience.Construction (
+    constructBalancedTx,
+
+    -- * Misc
+    TxInsExistError(..),
+    NotScriptLockedTxInsError(..),
+    notScriptLockedTxIns,
+    renderNotScriptLockedTxInsError,
+    renderTxInsExistError,
+    txinsExist,
```
Comment (james-iohk) [inline cardano-api/src/Cardano/Api/Convenience/Construction.hs:12]: `txInsExist` ?

Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Convenience/Construction.hs:12]: Open to naming suggestions: `txInsExistInUTxO`? Or perhaps I shouldn't export it in the first place

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 3]
```diff
@@ -665,26 +600,34 @@ txFeatureMismatch era feature =
...
+readScriptWitnessFiles _ [] = return []
+readScriptWitnessFiles era ((txin, mScriptWitnessFiles) : rest) = do
+  case  mScriptWitnessFiles of
+    Nothing -> do
+      remaining <- readScriptWitnessFiles era rest
+      return $ (txin, Nothing) : remaining
+    Just sWitFile -> do
+      sWit <- createScriptWitness era sWitFile
+      remaining <- readScriptWitnessFiles era rest
+      return $ (txin, Just sWit) : remaining
```
Comment (LudvikGalois) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:630]: Is there a reason to prefer explicit recursion over something like

```suggestion
readScriptWitnessFiles era = mapM go
  where
    go (txin, mScriptWitnessFiles) = do
      mSWit <- mapM (readScriptWitnessFiles era) mScriptWitnessFiles
      return (txin, mSWit)
```

File: cardano-api/src/Cardano/Api/Convenience/Construction.hs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+txinsExist :: [TxIn] -> UTxO era -> Either TxInsExistError ()
+txinsExist ins (UTxO utxo)
+  | null utxo = Left EmptyUTxO
+  | otherwise = do
+      let utxoIns = Map.keys utxo
+          occursInUtxo = [ txin | txin <- ins, txin `elem` utxoIns ]
+      if length occursInUtxo == length ins
+      then return ()
+      else Left . TxInsDoNotExist $ ins List.\\ ins `List.intersect` occursInUtxo
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/Convenience/Construction.hs:85]: Everything in `occursInUtxo` is drawn from `ins` to begin with, so intersecting with `ins` is identity (and therefore a waste of cycles)

```suggestion
      else Left . TxInsDoNotExist $ ins List.\\ occursInUtxo
```

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs (hunk 3 above)
Comment (erikd) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:630]: I agree! :+1:

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:630]: Refactored :+1:

File: cardano-api/src/Cardano/Api/Convenience/Construction.hs (hunk 4 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Convenience/Construction.hs:85]: good catch :+1:

**Review by LudvikGalois — APPROVED** [review]

**Review by erikd — APPROVED** [review]

**Comment by Jimbo4350** [comment]
bors merge

**Review by deepfire — APPROVED** [review]

**Comment by Jimbo4350** [comment]
bors merge

**Pull request merged**